File: sdks/python/src/graphrec_sdk/_retry.py
"""Retry policy.

The SDK retries only when doing so cannot duplicate a side effect:

* **Rate limiting** (``429 rate_limit_exceeded``) and **declared-retryable
  503s** are rejected before GraphRec changes anything, so every route may
  retry them. ``Retry-After`` / ``retry_after_seconds`` is honoured.
* **Connection failures before the request was sent** (DNS, refused, pool
  timeout) are always safe.
* **Ambiguous failures** (read timeouts, dropped connections, 502/504 from a
  proxy) are retried only for idempotent routes - reads, upserts, and writes
  deduplicated by ``event_id`` or ``Idempotency-Key``.

``quota_exceeded`` and every other 4xx are never retried.
"""

from __future__ import annotations

import random
from dataclasses import dataclass
from typing import Optional

import httpx

from ._constants import (
    DEFAULT_MAX_RETRIES,
    INITIAL_RETRY_DELAY,
    MAX_RETRY_AFTER_SECONDS,
    MAX_RETRY_DELAY,
)
from ._routes import Route
from .errors import APIStatusError

__all__ = ["RetryPolicy"]

_PRE_SEND_ERRORS = (httpx.ConnectError, httpx.ConnectTimeout, httpx.PoolTimeout)


@dataclass(frozen=True)
class RetryPolicy:
    max_retries: int = DEFAULT_MAX_RETRIES
    initial_delay: float = INITIAL_RETRY_DELAY
    max_delay: float = MAX_RETRY_DELAY
    #: A server-requested wait longer than this is surfaced as an error instead.
    max_retry_after: float = MAX_RETRY_AFTER_SECONDS
    jitter: float = 0.25

    def __post_init__(self) -> None:
        if self.max_retries < 0:
            raise ValueError("max_retries must be >= 0")

    def backoff(self, attempt: int) -> float:
        """Exponential backoff with jitter for zero-based ``attempt``."""

        base = min(self.initial_delay * (2.0**attempt), self.max_delay)
        spread = base * self.jitter
        return max(0.0, base + random.uniform(-spread, spread))

    def delay_for_status(
        self, error: APIStatusError, route: Route, attempt: int
    ) -> Optional[float]:
        """Seconds to wait before retrying ``error``, or ``None`` to give up."""

        if attempt >= self.max_retries:
            return None
        retry_after = error.retry_after_seconds
        if error.code == "quota_exceeded":
            return None
        rejected_before_processing = error.status_code == 429 or (
            error.status_code == 503 and error.retryable
        )
        gateway_failure = error.status_code in {502, 504} and route.idempotent
        if not (rejected_before_processing or gateway_failure):
            return None
        if retry_after is not None:
            if retry_after > self.max_retry_after:
                return None
            return retry_after
        return self.backoff(attempt)

    def delay_for_exception(
        self, exc: httpx.TransportError, route: Route, attempt: int
    ) -> Optional[float]:
        if attempt >= self.max_retries:
            return None
        if isinstance(exc, _PRE_SEND_ERRORS) or route.idempotent:
            return self.backoff(attempt)
        return None

File: sdks/typescript/src/retry.ts
/**
 * Retry policy.
 *
 * The SDK retries only when doing so cannot duplicate a side effect:
 *
 * - Rate limiting (`429 rate_limit_exceeded`) and declared-retryable 503s are
 *   rejected before GraphRec changes anything, so every route may retry them.
 *   `Retry-After` / `retry_after_seconds` is honoured.
 * - Connection failures before the request was sent are always safe.
 * - Ambiguous failures (timeouts, dropped connections, 502/504 from a proxy)
 *   are retried only for idempotent routes: reads, upserts, and writes
 *   deduplicated by `event_id` or `Idempotency-Key`.
 *
 * `quota_exceeded` and every other 4xx are never retried.
 */
import { DEFAULT_MAX_RETRIES, INITIAL_RETRY_DELAY_MS, MAX_RETRY_AFTER_SECONDS, MAX_RETRY_DELAY_MS } from "./constants.js";
import { APIStatusError, APITimeoutError } from "./errors.js";
import type { Route } from "./routes.js";

export interface RetryOptions {
  maxRetries?: number;
  initialDelayMs?: number;
  maxDelayMs?: number;
  /** A server-requested wait longer than this (seconds) is surfaced as an error instead. */
  maxRetryAfterSeconds?: number;
  jitter?: number;
}

export class RetryPolicy {
  readonly maxRetries: number;
  readonly initialDelayMs: number;
  readonly maxDelayMs: number;
  readonly maxRetryAfterSeconds: number;
  readonly jitter: number;

  constructor(options: RetryOptions = {}) {
    this.maxRetries = options.maxRetries ?? DEFAULT_MAX_RETRIES;
    this.initialDelayMs = options.initialDelayMs ?? INITIAL_RETRY_DELAY_MS;
    this.maxDelayMs = options.maxDelayMs ?? MAX_RETRY_DELAY_MS;
    this.maxRetryAfterSeconds = options.maxRetryAfterSeconds ?? MAX_RETRY_AFTER_SECONDS;
    this.jitter = options.jitter ?? 0.25;
    if (this.maxRetries < 0) throw new RangeError("maxRetries must be >= 0");
  }

  /** Exponential backoff with jitter, in milliseconds, for zero-based `attempt`. */
  backoffMs(attempt: number): number {
    const base = Math.min(this.initialDelayMs * 2 ** attempt, this.maxDelayMs);
    const spread = base * this.jitter;
    return Math.max(0, base + (Math.random() * 2 - 1) * spread);
  }

  /** Milliseconds to wait before retrying `error`, or `null` to give up. */
  delayForStatus(error: APIStatusError, route: Route, attempt: number): number | null {
    if (attempt >= this.maxRetries) return null;
    if (error.code === "quota_exceeded") return null;
    const rejectedBeforeProcessing = error.status === 429 || (error.status === 503 && error.retryable);
    const gatewayFailure = (error.status === 502 || error.status === 504) && route.idempotent;
    if (!rejectedBeforeProcessing && !gatewayFailure) return null;
    if (error.retryAfterSeconds !== undefined) {
      if (error.retryAfterSeconds > this.maxRetryAfterSeconds) return null;
      return error.retryAfterSeconds * 1000;
    }
    return this.backoffMs(attempt);
  }

  /**
   * Milliseconds to wait before retrying a transport failure, or `null`.
   * `fetch` cannot tell a pre-send connection failure from a dropped
   * connection, so a non-timeout transport error is treated as ambiguous and
   * retried only on idempotent routes; timeouts likewise.
   */
  delayForTransport(error: unknown, route: Route, attempt: number): number | null {
    if (attempt >= this.maxRetries) return null;
    if (error instanceof APITimeoutError) return route.idempotent ? this.backoffMs(attempt) : null;
    if (isConnectionRefused(error) || route.idempotent) return this.backoffMs(attempt);
    return null;
  }
}

/** Errors Node's fetch raises before any bytes were sent (safe to retry on every route). */
function isConnectionRefused(error: unknown): boolean {
  const cause = (error as { cause?: { code?: string } } | undefined)?.cause;
  const code = cause?.code ?? (error as { code?: string } | undefined)?.code;
  // ECONNRESET is deliberately absent: the request may already have been processed.
  return code === "ECONNREFUSED" || code === "ENOTFOUND" || code === "EAI_AGAIN";
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
